### PR TITLE
Integrate SplitExplicit01 code with OceanProblems code

### PR DESCRIPTION
**Pull request by blallen**
# Description

Makes it so the SplitExplicit01 code has shares the same problems as the rest of the Ocean code and provides defaults for each case. Also does some integration of the boundary condition handling across the different ocean models.

I have

- [x] Written and run all necessary tests with CLIMA by including `tests/runtests.jl`
- [x] Followed all necessary [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) and run `julia .dev/climaformat.jl .`
- [x] Updated the documentation to reflect changes from this PR.

# For review by CLIMA developers

- [x] There are no open pull requests for this already
- [x] CLIMA developers with relevant expertise have been assigned to review this submission
- [x] The code conforms to the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) and has consistent naming conventions. `julia .dev/format.jl` has been run in a separate commit.
- [x] This code does what it is technically intended to do (all numerics make sense physically and/or computationally)

### CODE REVIEW

**Comment by blallen** [comment]
bors try

**Comment by blallen** [comment]
bors try

**Comment by blallen** [comment]
bors try

**Comment by blallen** [comment]
bors try

**Comment by blallen** [comment]
bors try

**Comment by blallen** [comment]
> ## try
> Build succeeded:
> 
> * [test-os (windows-latest)](https://github.com/CliMA/ClimateMachine.jl/runs/1167191981)
> * [format](https://github.com/CliMA/ClimateMachine.jl/runs/1167191961)
> * [docs-build](https://github.com/CliMA/ClimateMachine.jl/runs/1167191962)
> * [buildkite/climatemachine-ci](https://buildkite.com/clima/climatemachine-ci/builds/414)
> * [test-os (macos-latest)](https://github.com/CliMA/ClimateMachine.jl/runs/1167191992)
> * [test-os (ubuntu-latest)](https://github.com/CliMA/ClimateMachine.jl/runs/1167191960)

@jm-c can you take a look at this PR? All the numerics check out, so it's just interface changes. you can still update the IC or the BC in each script if you want to, this just provides defaults.

**Comment by jm-c** [comment]
@blallen Will take a look and review it tomorrow (Mon). Was busy with other things.

**Review by jm-c — COMMENTED** [review]
@blallen It was not very clear to me where this new API was going to until I look at 
how you re-arrange the "ocean_gyre.jl" experiment and related set-up and tests. 
I think it's a nice way to structure the code ; and looking again at this set of changes
from a different perspective (more like a way towards similar same structure),
it make sense and am I up for this.
I am still not sure that it's the best time for these changes but this is more 
a matter of personal view.

Two general short comments:
1) there is not much comments in the code that would help to understand it;
  any addition would be helpful, not only for others, but even may be for you (after
  some time, anyone might forget why it's there ...).
2) the git commit comments/messages are not really reflecting all what has been changed,
  and this make it harder to follow.
  For instance, some of the changes are there to "finish" applying PR 1554 (removing Qvec
  in some split-explicit set-up), but only "diff" will let me know.
  Also in SplitExplicit01/OceanBoundaryConditions.jl, the ordering in argument list for 
  "ocean_boundary_state" have been switched, but I have not found relevant comments.
  Nothing to change here but may be next time ...
  
Little things:   
a) I would prefer to keep in some commented lines
   e.g., in test/Ocean/SplitExplicit/simple_box_ivd.jl,
   lines 188-190 and also lines 261-264 & 271-272, and 433-434 ;
   Because this is a nice way to remember how to make simple modification to 
   change some parameter value or to add more variable to StateCheck (I used that 
   more than once when debugging).
b) Would need your help in bringing up-to-date some developments from my side
   because of (1) and mainly (2), it's not obvious to follow the changes without 
   spending too much time. And other development need also to move forward.

File: src/Ocean/OceanBC.jl [hunk 1]
```diff
@@ -86,3 +87,89 @@ Prescribe the net inward temperature flux across the boundary by `flux`,
...
 """
 struct TemperatureFlux <: TemperatureBC end
+
+"""
+    ocean_boundary_state!(nf, boundaries::Tuple, ::HBModel,
+                          Q⁺, A⁺, n, Q⁻, A⁻, bctype)
+applies boundary conditions for the first-order and gradient fluxes
+dispatches to a function in OceanBoundaryConditions.jl based on bytype defined by a problem such as SimpleBoxProblem.jl
+"""
+@generated function ocean_boundary_state!(
```
Comment (sandreza) [inline src/Ocean/OceanBC.jl:97]: This is starting to get fancy

**Review by sandreza — APPROVED** [review]

File: src/Ocean/OceanBC.jl (hunk 1 above)
Comment (blallen) [reply src/Ocean/OceanBC.jl:97]: the `@generated` magic is thanks to @simonbyrne, I've just stolen the idea

**Comment by christophernhill** [comment]
@blallen and @sandreza are you sure you want to go the ```@generated``` route?

it may have some downsides e.g. 

https://discourse.julialang.org/t/why-not-use-generated/29531/4

and can introduce readability issues. 
It can sometimes be useful to be able to easily read the code for things like bc's,  which this could make a little trickier for some folks. 

Not sure what right answer is, so more musing out loud than anything!

**Comment by blallen** [comment]
> @blallen and @sandreza are you sure you want to go the `@generated` route?
> 
> it may have some downsides e.g.
> 
> https://discourse.julialang.org/t/why-not-use-generated/29531/4
> 
> and can introduce readability issues.
> It can sometimes be useful to be able to easily read the code for things like bc's, which this could make a little trickier for some folks.
> 
> Not sure what right answer is, so more musing out loud than anything!

We are moving towards this interface project wide: see https://github.com/CliMA/ClimateMachine.jl/pull/1522
